File: multi-processor-programming-study/exercises/cpp/10_queues/exercise01_concurrent_queues.cpp
/**
 * Exercise: Concurrent Queues (C++20)
 *
 * CONCEPT: Lock-free and blocking concurrent queues
 *
 * Queue implementations:
 * - Bounded Blocking: Fixed capacity with condition variables
 * - Lock-Free: Michael-Scott queue using CAS
 *
 * LEARNING OBJECTIVES:
 * - Implement blocking queue with condition variables
 * - Implement Michael-Scott lock-free queue
 * - Handle the ABA problem
 * - Understand producer-consumer patterns
 */

#include <iostream>
#include <thread>
#include <vector>
#include <atomic>
#include <mutex>
#include <condition_variable>
#include <optional>
#include <chrono>
#include <iomanip>

namespace multiprocessor::queues {

/**
 * TODO: Implement Bounded Blocking Queue
 *
 * Fixed-size array-based queue with blocking
 */
template <typename T>
class BoundedQueue {
private:
    std::vector<T> items;
    int head = 0;
    int tail = 0;
    int size = 0;
    const int capacity;

    // TODO: Add mutex and condition variables
    // HINT: std::mutex, std::condition_variable for not_full and not_empty

public:
    explicit BoundedQueue(int cap) : items(cap), capacity(cap) {
        // TODO: Initialize
    }

    /**
     * TODO: Implement blocking enqueue
     *
     * Wait while full, then add item
     */
    void enqueue(const T& item) {
        // TODO: Lock mutex
        // TODO: Wait on not_full while size == capacity
        // TODO: Add item to items[tail]
        // TODO: Update tail and size
        // TODO: Notify not_empty
    }

    /**
     * TODO: Implement blocking dequeue
     *
     * Wait while empty, then remove item
     */
    T dequeue() {
        // TODO: Lock mutex
        // TODO: Wait on not_empty while size == 0
        // TODO: Get item from items[head]
        // TODO: Update head and size
        // TODO: Notify not_full
        // TODO: Return item
        return T{};
    }

    int get_size() const {
        return size;
    }
};

/**
 * TODO: Implement Michael-Scott Lock-Free Queue
 *
 * Classic lock-free queue using CAS
 * Handles concurrent enqueue and dequeue operations
 */
template <typename T>
class LockFreeQueue {
private:
    struct Node {
        T value;
        std::atomic<Node*> next;

        Node() : next(nullptr) {}
        explicit Node(const T& val) : value(val), next(nullptr) {}
    };

    // TODO: Add std::atomic<Node*> for head and tail

public:
    LockFreeQueue() {
        // TODO: Initialize with dummy node
        // HINT: Both head and tail point to dummy initially
    }

    ~LockFreeQueue() {
        // TODO: Clean up remaining nodes
    }

    /**
     * TODO: Implement lock-free enqueue (Michael-Scott algorithm)
     *
     * 1. Create new node
     * 2. Loop:
     *    a. Read tail and next
     *    b. Check consistency (tail hasn't changed)
     *    c. If tail.next is null, try CAS it to new node
     *    d. If CAS succeeds, try to swing tail forward
     *    e. If tail.next not null, help other thread by swinging tail
     */
    void enqueue(const T& item) {
        Node* node = new Node(item);
        // TODO: Implement Michael-Scott enqueue
    }

    /**
     * TODO: Implement lock-free dequeue
     *
     * 1. Loop:
     *    a. Read head, tail, and first real node
     *    b. Check consistency
     *    c. If queue empty (head == tail), return nullopt
     *    d. If tail falling behind, help advance it
     *    e. Read value before CAS (important!)
     *    f. Try to swing head forward with CAS
     */
    std::optional<T> dequeue() {
        // TODO: Implement Michael-Scott dequeue
        // HINT: Skip dummy node, return first real node's value
        return std::nullopt;
    }
};

/**
 * Performance testing
 */
class PerformanceTest {
public:
    template <typename Queue>
    static void test_queue(const std::string& name, Queue& queue,
                          int num_producers, int num_consumers,
                          int items_per_producer) {
        std::vector<std::thread> threads;

        auto start_time = std::chrono::high_resolution_clock::now();

        // Producers
        for (int i = 0; i < num_producers; ++i) {
            threads.emplace_back([&queue, i, items_per_producer]() {
                for (int j = 0; j < items_per_producer; ++j) {
                    queue.enqueue(i * items_per_producer + j);
                }
            });
        }

        // Consumers
        std::atomic<int> consumed{0};
        for (int i = 0; i < num_consumers; ++i) {
            threads.emplace_back([&queue, &consumed, items_per_producer]() {
                for (int j = 0; j < items_per_producer; ++j) {
                    auto item = queue.dequeue();
                    if constexpr (std::is_same_v<Queue, LockFreeQueue<int>>) {
                        if (item.has_value()) {
                            consumed.fetch_add(1);
                        }
                    } else {
                        consumed.fetch_add(1);
                    }
                }
            });
        }

        for (auto& t : threads) {
            t.join();
        }

        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
            end_time - start_time).count();

        std::cout << std::left << std::setw(25) << name << ": "
                  << std::right << std::setw(6) << duration << "ms"
                  << " (consumed=" << consumed.load() << ")\n";
    }
};

void test_bounded_queue() {
    std::cout << "Testing Bounded Queue:\n";

    BoundedQueue<int> queue(100);
    const int num_producers = 5;
    const int num_consumers = 5;
    const int items_per_producer = 1000;

    std::vector<std::thread> threads;

    // Producers
    for (int i = 0; i < num_producers; ++i) {
        threads.emplace_back([&queue, i, items_per_producer]() {
            for (int j = 0; j < items_per_producer; ++j) {
                queue.enqueue(i * items_per_producer + j);
            }
        });
    }

    // Consumers
    std::atomic<int> consumed{0};
    for (int i = 0; i < num_consumers; ++i) {
        threads.emplace_back([&queue, &consumed, items_per_producer]() {
            for (int j = 0; j < items_per_producer; ++j) {
                queue.dequeue();
                consumed.fetch_add(1);
            }
        });
    }

    for (auto& t : threads) {
        t.join();
    }

    int expected = num_producers * items_per_producer;
    int actual = consumed.load();

    std::cout << "Produced: " << expected << "\n";
    std::cout << "Consumed: " << actual << "\n";
    std::cout << (actual == expected ? "✅ PASS\n\n" : "❌ FAIL\n\n");
}

void test_lock_free_queue() {
    std::cout << "Testing Lock-Free Queue:\n";

    LockFreeQueue<int> queue;
    const int num_threads = 10;
    const int items_per_thread = 1000;

    // Enqueue phase
    std::vector<std::thread> enqueuers;
    for (int i = 0; i < num_threads; ++i) {
        enqueuers.emplace_back([&queue, i, items_per_thread]() {
            for (int j = 0; j < items_per_thread; ++j) {
                queue.enqueue(i * items_per_thread + j);
            }
        });
    }
    for (auto& t : enqueuers) t.join();

    // Dequeue phase
    std::atomic<int> dequeue_count{0};
    std::vector<std::thread> dequeuers;
    for (int i = 0; i < num_threads; ++i) {
        dequeuers.emplace_back([&queue, &dequeue_count, items_per_thread]() {
            for (int j = 0; j < items_per_thread; ++j) {
                auto item = queue.dequeue();
                if (item.has_value()) {
                    dequeue_count.fetch_add(1);
                }
            }
        });
    }
    for (auto& t : dequeuers) t.join();

    int expected = num_threads * items_per_thread;
    int actual = dequeue_count.load();

    std::cout << "Expected: " << expected << "\n";
    std::cout << "Dequeued: " << actual << "\n";
    std::cout << (actual == expected ? "✅ PASS\n\n" : "❌ FAIL\n\n");
}

void demonstrate_aba_problem() {
    std::cout << "=== ABA Problem in Queues ===\n";
    std::cout << "Problem: Node reuse can cause ABA issues\n";
    std::cout << "Thread 1: Reads node A\n";
    std::cout << "Thread 2: Dequeues A, dequeues B, enqueues A (reused)\n";
    std::cout << "Thread 1: CAS succeeds, but A is different!\n\n";
    std::cout << "Solutions:\n";
    std::cout << "- Don't reuse nodes immediately\n";
    std::cout << "- Use version numbers/stamps\n";
    std::cout << "- Use hazard pointers\n";
    std::cout << "- Use epoch-based reclamation\n\n";
}

} // namespace multiprocessor::queues

int main() {
    using namespace multiprocessor::queues;

    std::cout << "=== Concurrent Queues Test (C++) ===\n\n";

    test_bounded_queue();
    test_lock_free_queue();
    demonstrate_aba_problem();

    return 0;
}

File: multi-processor-programming-study/exercises/java/src/main/java/com/multiprocessor/queues/Exercise01_ConcurrentQueues.java
package com.multiprocessor.queues;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Exercise: Concurrent Queues
 *
 * CONCEPT: Producer-consumer queues with different synchronization strategies
 *
 * Queue types:
 * - Bounded Blocking: Fixed capacity, threads block when full/empty
 * - Unbounded Lock-Based: Dynamic capacity, uses locks
 * - Lock-Free: Michael-Scott queue using CAS
 * - Array-Based: Circular array, cache-friendly
 *
 * LEARNING OBJECTIVES:
 * - Implement blocking and non-blocking queues
 * - Handle the ABA problem
 * - Master the Michael-Scott algorithm
 */
public class Exercise01_ConcurrentQueues {

    /**
     * TODO: Implement Bounded Blocking Queue
     *
     * Fixed-size array-based queue with blocking
     * Threads block when queue is full (enqueue) or empty (dequeue)
     */
    public static class BoundedQueue<T> {
        private final T[] items;
        private int head = 0;
        private int tail = 0;
        private int size = 0;
        // TODO: Add lock and conditions
        // HINT: Use Lock, Condition for notFull and notEmpty

        @SuppressWarnings("unchecked")
        public BoundedQueue(int capacity) {
            items = (T[]) new Object[capacity];
            // TODO: Initialize lock and conditions
        }

        /**
         * TODO: Implement blocking enqueue
         *
         * Wait while full, then add item
         *
         * @param item item to enqueue
         * @throws InterruptedException if interrupted while waiting
         */
        public void enqueue(T item) throws InterruptedException {
            // TODO: Implement
            // Lock
            // While full: wait on notFull condition
            // Add item
            // Update tail and size
            // Signal notEmpty
            // Unlock
        }

        /**
         * TODO: Implement blocking dequeue
         *
         * Wait while empty, then remove item
         *
         * @return dequeued item
         * @throws InterruptedException if interrupted while waiting
         */
        public T dequeue() throws InterruptedException {
            // TODO: Implement
            // Lock
            // While empty: wait on notEmpty condition
            // Remove item
            // Update head and size
            // Signal notFull
            // Unlock
            return null;
        }

        public int size() {
            return size;
        }
    }

    /**
     * TODO: Implement Unbounded Lock-Based Queue
     *
     * Uses linked nodes, separate locks for head and tail
     * Better concurrency than single lock
     */
    public static class UnboundedQueue<T> {
        private class Node {
            T value;
            Node next;

            Node(T value) {
                this.value = value;
            }
        }

        private Node head;
        private Node tail;
        // TODO: Add separate locks for head and tail
        // HINT: enqLock for tail operations, deqLock for head operations

        public UnboundedQueue() {
            Node sentinel = new Node(null);
            head = sentinel;
            tail = sentinel;
            // TODO: Initialize locks
        }

        /**
         * TODO: Implement enqueue with tail lock
         *
         * Lock tail, add node, unlock
         */
        public void enqueue(T item) {
            Node node = new Node(item);
            // TODO: Lock tail
            // TODO: Add node
            // TODO: Update tail
            // TODO: Unlock
        }

        /**
         * TODO: Implement dequeue with head lock
         *
         * Lock head, remove node, unlock
         * Return null if empty
         */
        public T dequeue() {
            // TODO: Lock head
            // TODO: Check if queue is empty (head == tail)
            // TODO: If not empty, remove first real node
            // TODO: Update head
            // TODO: Unlock
            return null;
        }
    }

    /**
     * TODO: Implement Michael-Scott Lock-Free Queue
     *
     * Classic lock-free queue using CAS
     * Handles concurrent enqueue and dequeue operations
     */
    public static class LockFreeQueue<T> {
        private class Node {
            final T value;
            final AtomicReference<Node> next;

            Node(T value) {
                this.value = value;
                this.next = new AtomicReference<>(null);
            }
        }

        private final AtomicReference<Node> head;
        private final AtomicReference<Node> tail;

        public LockFreeQueue() {
            Node sentinel = new Node(null);
            head = new AtomicReference<>(sentinel);
            tail = new AtomicReference<>(sentinel);
        }

        /**
         * TODO: Implement lock-free enqueue
         *
         * Michael-Scott algorithm:
         * 1. Create new node
         * 2. Loop:
         *    a. Read tail and next
         *    b. If tail.next is null, try CAS it to new node
         *    c. If CAS succeeds, try to swing tail forward
         *    d. If tail.next not null, help other thread by swinging tail
         */
        public void enqueue(T item) {
            Node node = new Node(item);
            while (true) {
                Node last = tail.get();
                Node next = last.next.get();

                if (last == tail.get()) { // Consistency check
                    if (next == null) {
                        // TODO: Try to link new node
                        if (last.next.compareAndSet(next, node)) {
                            // TODO: Try to swing tail
                            tail.compareAndSet(last, node);
                            return;
                        }
                    } else {
                        // TODO: Help advance tail
                        tail.compareAndSet(last, next);
                    }
                }
            }
        }

        /**
         * TODO: Implement lock-free dequeue
         *
         * 1. Loop:
         *    a. Read head, tail, and first real node
         *    b. Check consistency
         *    c. If queue empty (head == tail), return null
         *    d. If tail falling behind, help advance it
         *    e. Try to swing head forward with CAS
         */
        public T dequeue() {
            while (true) {
                Node first = head.get();
                Node last = tail.get();
                Node next = first.next.get();

                if (first == head.get()) { // Consistency check
                    if (first == last) {
                        // Queue empty or tail falling behind
                        if (next == null) {
                            return null; // Empty
                        }
                        // TODO: Help advance tail
                        tail.compareAndSet(last, next);
                    } else {
                        // TODO: Read value before CAS
                        T value = next.value;
                        // TODO: Try to swing head forward
                        if (head.compareAndSet(first, next)) {
                            return value;
                        }
                    }
                }
            }
        }
    }

    /**
     * Demonstration of the ABA problem
     */
    public static class ABADemo {
        /**
         * The ABA problem:
         * - Thread 1 reads A
         * - Thread 2 changes A->B->A
         * - Thread 1's CAS succeeds but state changed
         *
         * Solutions:
         * - Version numbers (stamps)
         * - Hazard pointers
         * - Epoch-based reclamation
         */
        public static void demonstrate() {
            System.out.println("=== ABA Problem in Queues ===");
            System.out.println("Problem: Node reuse can cause ABA issues");
            System.out.println("Thread 1: Reads node A");
            System.out.println("Thread 2: Dequeues A, dequeues B, enqueues A (reused)");
            System.out.println("Thread 1: CAS succeeds, but A is different!");
            System.out.println();
            System.out.println("Solutions:");
            System.out.println("- Don't reuse nodes immediately");
            System.out.println("- Use version numbers with AtomicStampedReference");
            System.out.println("- Use hazard pointers or epoch-based reclamation");
            System.out.println();
        }
    }

    /**
     * Performance comparison
     */
    static class PerformanceTest {
        static void testQueue(String name, Runnable test) throws InterruptedException {
            long startTime = System.nanoTime();
            test.run();
            long endTime = System.nanoTime();
            long duration = (endTime - startTime) / 1_000_000; // ms
            System.out.printf("%20s: %6dms%n", name, duration);
        }
    }

    /**
     * Demonstration and testing
     */
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Concurrent Queues Test ===\n");

        testBoundedQueue();
        testUnboundedQueue();
        testLockFreeQueue();
        ABADemo.demonstrate();
    }

    private static void testBoundedQueue() throws InterruptedException {
        System.out.println("Testing Bounded Queue:");

        BoundedQueue<Integer> queue = new BoundedQueue<>(100);
        int numProducers = 5;
        int numConsumers = 5;
        int itemsPerProducer = 1000;

        Thread[] producers = new Thread[numProducers];
        for (int i = 0; i < numProducers; i++) {
            final int id = i;
            producers[i] = new Thread(() -> {
                try {
                    for (int j = 0; j < itemsPerProducer; j++) {
                        queue.enqueue(id * itemsPerProducer + j);
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
            producers[i].start();
        }

        AtomicInteger consumedCount = new AtomicInteger(0);
        Thread[] consumers = new Thread[numConsumers];
        for (int i = 0; i < numConsumers; i++) {
            consumers[i] = new Thread(() -> {
                try {
                    for (int j = 0; j < itemsPerProducer; j++) {
                        Integer item = queue.dequeue();
                        if (item != null) {
                            consumedCount.incrementAndGet();
                        }
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
            consumers[i].start();
        }

        for (Thread t : producers) t.join();
        for (Thread t : consumers) t.join();

        System.out.println("Produced: " + (numProducers * itemsPerProducer));
        System.out.println("Consumed: " + consumedCount.get());
        System.out.println(consumedCount.get() == numProducers * itemsPerProducer ?
                          "✅ PASS\n" : "❌ FAIL\n");
    }

    private static void testUnboundedQueue() throws InterruptedException {
        System.out.println("Testing Unbounded Queue:");

        UnboundedQueue<Integer> queue = new UnboundedQueue<>();
        int numThreads = 10;
        int itemsPerThread = 1000;

        // Enqueue phase
        Thread[] enqueuers = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            final int id = i;
            enqueuers[i] = new Thread(() -> {
                for (int j = 0; j < itemsPerThread; j++) {
                    queue.enqueue(id * itemsPerThread + j);
                }
            });
            enqueuers[i].start();
        }

        for (Thread t : enqueuers) t.join();

        // Dequeue phase
        AtomicInteger dequeueCount = new AtomicInteger(0);
        Thread[] dequeuers = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            dequeuers[i] = new Thread(() -> {
                for (int j = 0; j < itemsPerThread; j++) {
                    Integer item = queue.dequeue();
                    if (item != null) {
                        dequeueCount.incrementAndGet();
                    }
                }
            });
            dequeuers[i].start();
        }

        for (Thread t : dequeuers) t.join();

        System.out.println("Expected: " + (numThreads * itemsPerThread));
        System.out.println("Dequeued: " + dequeueCount.get());
        System.out.println(dequeueCount.get() == numThreads * itemsPerThread ?
                          "✅ PASS\n" : "❌ FAIL\n");
    }

    private static void testLockFreeQueue() throws InterruptedException {
        System.out.println("Testing Lock-Free Queue:");

        LockFreeQueue<Integer> queue = new LockFreeQueue<>();
        int numThreads = 10;
        int itemsPerThread = 1000;

        // Mixed operations
        Thread[] threads = new Thread[numThreads * 2];

        // Half enqueue
        for (int i = 0; i < numThreads; i++) {
            final int id = i;
            threads[i] = new Thread(() -> {
                for (int j = 0; j < itemsPerThread; j++) {
                    queue.enqueue(id * itemsPerThread + j);
                }
            });
        }

        // Half dequeue
        AtomicInteger dequeueCount = new AtomicInteger(0);
        for (int i = 0; i < numThreads; i++) {
            threads[numThreads + i] = new Thread(() -> {
                for (int j = 0; j < itemsPerThread; j++) {
                    Integer item = queue.dequeue();
                    if (item != null) {
                        dequeueCount.incrementAndGet();
                    }
                }
            });
        }

        for (Thread t : threads) t.start();
        for (Thread t : threads) t.join();

        System.out.println("Enqueued: " + (numThreads * itemsPerThread));
        System.out.println("Dequeued: " + dequeueCount.get());
        System.out.println(dequeueCount.get() <= numThreads * itemsPerThread ?
                          "✅ PASS (some items may remain in queue)\n" : "❌ FAIL\n");
    }
}
